buffer/src/main/java/io/netty/buffer/PoolChunk.java

```diff
@@ -110,239 +212,274 @@ int usage() {
+                subpage.init(head, normCapacity);
+            }
+            return subpage.allocate();
         }
-        return -1;
     }
 
+    /**
+     * Free a subpage or a run of pages
+     * When a subpage is freed from PoolSubpage, it might be added back to subpage pool of the owning PoolArena
+     * If the subpage pool in PoolArena has at least one other PoolSubpage of given elemSize, we can
+     * completely free the owning Page so it is available for subsequent allocations
+     *
+     * @param handle handle to free
+     */
     void free(long handle) {
-        int memoryMapIdx = (int) handle;
-        int bitmapIdx = (int) (handle >>> 32);
+        int memoryMapIdx = memoryMapIdx(handle);
+        int bitmapIdx = bitmapIdx(handle);
 
-        int val = memoryMap[memoryMapIdx];
-        int state = val & 3;
-        if (state == ST_ALLOCATED_SUBPAGE) {
-            assert bitmapIdx != 0;
+        if (bitmapIdx != 0) { // free a subpage
             PoolSubpage<T> subpage = subpages[subpageIdx(memoryMapIdx)];
             assert subpage != null && subpage.doNotDestroy;
-            if (subpage.free(bitmapIdx & 0x3FFFFFFF)) {
-                return;
-            }
-        } else {
-            assert state == ST_ALLOCATED : "state: " + state;
-            assert bitmapIdx == 0;
-        }
-
-        freeBytes += runLength(val);
-
-        for (;;) {
-            //noinspection PointlessBitwiseExpression
-            memoryMap[memoryMapIdx] = val & ~3 | ST_UNUSED;
-            if (memoryMapIdx == 1) {
-                assert freeBytes == chunkSize;
-                return;
-            }
 
-            if ((memoryMap[siblingIdx(memoryMapIdx)] & 3) != ST_UNUSED) {
-                break;
+            // Obtain the head of the PoolSubPage pool that is owned by the PoolArena and synchronize on it.
+            // This is need as we may add it back and so alter the linked-list structure.
+            PoolSubpage<T> head = arena.findSubpagePoolHead(subpage.elemSize);
+            synchronized (head) {
+                if (subpage.free(head, bitmapIdx & 0x3FFFFFFF)) {
+                    return;
+                }
             }
-
-            memoryMapIdx = parentIdx(memoryMapIdx);
-            val = memoryMap[memoryMapIdx];
         }
+        freeBytes += runLength(memoryMapIdx);
+        setValue(memoryMapIdx, depth(memoryMapIdx));
+        updateParentsFree(memoryMapIdx);
     }
 
     void initBuf(PooledByteBuf<T> buf, long handle, int reqCapacity) {
-        int memoryMapIdx = (int) handle;
-        int bitmapIdx = (int) (handle >>> 32);
+        int memoryMapIdx = memoryMapIdx(handle);
+        int bitmapIdx = bitmapIdx(handle);
         if (bitmapIdx == 0) {
-            int val = memoryMap[memoryMapIdx];
-            assert (val & 3) == ST_ALLOCATED : String.valueOf(val & 3);
-            buf.init(this, handle, runOffset(val), reqCapacity, runLength(val));
+            byte val = value(memoryMapIdx);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this useless assignment to local variable "val". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1854)